File: 2/main.py
f = open('2/input.txt')

ROCK = 'rock'
PAPER = 'paper'
SCISSORS = 'scissors'

WIN = 'Z'
DRAW = 'Y'
LOSE = 'X'


def to_tool(char: str):
    if char in ['A', 'X']:
        return ROCK
    if char in ['B', 'Y']:
        return PAPER
    if char in ['C', 'Z']:
        return SCISSORS


def battle(p1, p2) -> int:
    if p1 == p2:
        return 1
    if p1 == ROCK:
        if p2 == PAPER:
            return 2
        else:
            return 0
    if p1 == PAPER:
        if p2 == SCISSORS:
            return 2
        else:
            return 0
    if p1 == SCISSORS:
        if p2 == ROCK:
            return 2
        else:
            return 0

def tool_score(tool):
    if tool == ROCK:
        return 1
    elif tool == PAPER:
        return 2
    else:
        return 3


def play(opponent: str, me:str):
    score = tool_score(me) + battle(opponent, me) * 3
    return score

nemesis_map = {ROCK: PAPER, PAPER: SCISSORS, SCISSORS: ROCK}

def find_tool(opponent: str, outcome:str):
    if outcome == DRAW:
        return opponent
    if outcome == LOSE:
        return nemesis_map[nemesis_map[opponent]]
    else:
        return nemesis_map[opponent]


def play2(opponent: str, outcome:str):
    my_tool = find_tool(opponent, outcome)
    score = tool_score(my_tool) + battle(opponent, my_tool) * 3
    return score


scores = []
scores2 = []
for l in f.readlines():
    scores.append(play(to_tool(l[0]), to_tool(l[2])))
    scores2.append(play2(to_tool(l[0]), l[2]))

print(sum(scores))
print(sum(scores2))


f.close()

File: 2/src/main.rs
use std::collections::HashMap;
use std::env;
use std::fs;

#[derive(Debug)]
#[derive(PartialEq)]
#[derive(Eq)]
#[derive(Hash)]
enum Tool {
    Rock,
    Paper,
    Scissors,
}
impl Tool {
    fn from_str(char: &str) -> Self{
        return match char {
                "A"|"X" => Tool::Rock,
                "B"|"Y" => Tool::Paper,
                _ => Tool::Scissors,
        };
 }
}
fn tool_score(tool: &Tool) -> i32{
    return match tool {
        Tool::Rock => 1,
        Tool::Paper => 2,
        Tool::Scissors => 3,
    };
}

fn nemesis_map(tool: &Tool) -> Tool {
    return match tool {
        Tool::Rock => Tool::Paper,
        Tool::Paper => Tool::Scissors,
        Tool::Scissors => Tool::Rock,
    }
}


fn battle(p1: &Tool, p2: &Tool) -> i32 {
    let nemesis_map = HashMap::from([
        (Tool::Rock, Tool::Paper),
        (Tool::Paper, Tool::Scissors),
        (Tool::Scissors, Tool::Rock),
    ]);
    if p1 == p2 {
        return 0;
    } 
    else if *p1 == nemesis_map[&p2] {
        return -1; 
    }
    else{
        return 1;
    }
}

#[derive(Debug)]
enum Outcome {
    Win,
    Draw,
    Lose,
}
impl Outcome {
    fn from_str(char: &str) -> Self{
        return match char {
                "X" => Outcome::Lose,
                "Y" => Outcome::Draw,
                _ => Outcome::Win,
        }
    }
}

fn find_tool(opponent: &Tool, outcome: &Outcome) -> Tool {
  return match outcome  {
      Outcome::Draw => match opponent {
        Tool::Paper => Tool::Paper,
        Tool::Rock => Tool::Rock,
        Tool::Scissors => Tool::Scissors,
      },
      Outcome::Lose => nemesis_map(&nemesis_map(opponent)),
      Outcome::Win => nemesis_map(opponent),
  }

}

fn play(opponent: &Tool, me: &Tool) -> i32 {
    let score = tool_score(me) + (battle(opponent, me) + 1)*3;
    return score;
}

fn play2(opponent: &Tool, outcome: &Outcome) -> i32 {
    let my_tool = find_tool(opponent, outcome);
    let score = tool_score(&my_tool) + (battle(opponent, &my_tool) + 1)*3;
    return score;
}

fn main() {
    let mut current_dir = env::current_dir().unwrap().to_str().unwrap().to_owned();
    let today = "/2";
    if !current_dir.ends_with(today) {
        current_dir += today
    }
    let contents = fs::read_to_string("input.txt").expect("couldn't read file");
    let lines = contents.lines();
    let mut first: Vec<i32> = vec![];
    let mut second: Vec<i32> = vec![];

    for l in lines.into_iter() {
        let tool = Tool::from_str(l.chars().nth(0).unwrap().to_string().as_str());
        let second_tool = Tool::from_str(l.chars().nth(2).unwrap().to_string().as_str());
        let outcome = Outcome::from_str(l.chars().nth(2).unwrap().to_string().as_str());
        first.extend([play(&tool, &second_tool)]);
        second.extend([play2(&tool, &outcome)]);

        println!("{:?}", tool);
        println!("{:?}", second_tool);
        println!("{:?}", outcome);
    }
    println!("{:?}", first);
    let sum: i32 = first.iter().sum();
    let sum2: i32 = second.iter().sum();
    println!("{sum}");
    println!("{sum2}");
}
